lr_face/losses.py

```diff
@@ -0,0 +1,61 @@
+import tensorflow as tf
+from tensorflow.keras.losses import Loss
+
+
+class TripletLoss(Loss):
+    def __init__(self, alpha: float, **kwargs):
+        """
+        Instantiate a triplet loss instance.
+
+        Arguments:
+            alpha: A margin between positive and negative pairs. Mathematically
+                for the loss to be 0, the following must hold:
+
+                    `dist(anchor, negative) - dist(anchor, positive) > alpha`
+
+                Where `dist(a, b)` is a distance measure between two embeddings
+                `a` and `b`. When `alpha` is 0, the loss can theoretically be
+                minimized by mapping every image to the same embedding.
+        """
+        super().__init__(**kwargs)
+        self.alpha = alpha
+
+    def call(self, y_true, y_pred):
+        """
+        Keras implementation of the triplet loss.
+
+        See https://en.wikipedia.org/wiki/Triplet_loss for more info.
+
+        Arguments:
+            y_true: Ignored, but required due to `Loss.call()` interface.
+            y_pred: A 3D tensor of shape `(batch_size, 3, embedding_size)`.
+                It contains the embeddings for each combination of anchor,
+                positive and negative image in the batch, respectively, which
+                is represented by the last axis of size 3. Make sure that each
```

Comment:
je voert hier de l2 normalisatie niet uit, en checkt er ook niet op? zou dat niet moeten? Waar doen we het wel?

Reply:
Erop checken in de loss vind ik wel een goed idee. Het uitvoeren ervan vind ik een verantwoordelijkheid van het model (bij `VGGFace` gebeurt dit nu al).

Reply:
Misschien is het beter om hier aparte unit tests voor te schrijven ipv in de triplet loss te testen. In principe is het namelijk niet iets wat de triplet loss per se zou moeten afdwingen.

Reply:
nee maar omdat je specifiek de euclidische afstand neemt ben je er wel gevoelig voor. Zou het niet raar vinden iig een warning te krijgen

Reply:
Ik heb het nu zo gemaakt dat je aan de `__init__()` van `TripletLoss` een boolean mee kunt geven of je wil afdwingen dat de input genormaliseerd is. Deze staat standaard op `True` (dus er wordt standaard wel gecontroleerd). Als blijkt dat de input niet genormaliseerd is, wordt er een exception geraised.